File: src/lib/bandsintown.ts
import { cappedSet } from "./boundedCache";

const APP_ID = "js_app_id";
const BASE_URL = "https://rest.bandsintown.com";

export interface BandsintownEvent {
  datetime: string;
  venueName: string;
  venueCity: string;
  venueRegion: string;
  venueCountry: string;
  url: string;
  lineup: string[];
}

const MAX_CACHE_ENTRIES = 2000;
const cache = new Map<string, BandsintownEvent[]>();
const inflight = new Map<string, Promise<BandsintownEvent[]>>();

function setCached(key: string, events: BandsintownEvent[]): void {
  cappedSet(cache, key, events, MAX_CACHE_ENTRIES);
}

function cacheKey(name: string): string {
  return name.trim().toLowerCase();
}

export async function fetchBandsintownEvents(artistName: string): Promise<BandsintownEvent[]> {
  const key = cacheKey(artistName);
  if (!key) return [];
  const hit = cache.get(key);
  if (hit !== undefined) return hit;
  const pending = inflight.get(key);
  if (pending) return pending;

  const promise = (async () => {
    try {
      const encoded = encodeURIComponent(artistName.trim());
      const res = await fetch(`${BASE_URL}/artists/${encoded}/events?app_id=${APP_ID}`);
      if (!res.ok) {
        setCached(key, []);
        return [];
      }
      const raw = (await res.json()) as Array<Record<string, unknown>>;
      if (!Array.isArray(raw)) {
        setCached(key, []);
        return [];
      }
      const events: BandsintownEvent[] = raw.slice(0, 20).map((item) => {
        const venue = (item.venue ?? {}) as Record<string, unknown>;
        const lineup = Array.isArray(item.lineup)
          ? (item.lineup as unknown[]).filter((s): s is string => typeof s === "string")
          : [];
        return {
          datetime: typeof item.datetime === "string" ? item.datetime : "",
          venueName: typeof venue.name === "string" ? venue.name : "",
          venueCity: typeof venue.city === "string" ? venue.city : "",
          venueRegion: typeof venue.region === "string" ? venue.region : "",
          venueCountry: typeof venue.country === "string" ? venue.country : "",
          url: typeof item.url === "string" ? item.url : "",
          lineup,
        };
      });
      setCached(key, events);
      return events;
    } catch {
      setCached(key, []);
      return [];
    } finally {
      inflight.delete(key);
    }
  })();

  inflight.set(key, promise);
  return promise;
}

File: src-tauri/src/bandsintown.rs
const APP_ID: &str = "js_app_id";

#[derive(serde::Serialize, Default)]
pub struct BandsintownEvent {
    pub datetime: String,
    pub venue_name: String,
    pub venue_city: String,
    pub venue_region: String,
    pub venue_country: String,
    pub url: String,
    pub lineup: Vec<String>,
}

/// Fetch upcoming Bandsintown events for an artist. Returns empty on any error.
#[tauri::command]
pub async fn fetch_bandsintown_events(artist_name: String) -> Result<Vec<BandsintownEvent>, String> {
    let trimmed = artist_name.trim();
    if trimmed.is_empty() {
        return Ok(vec![]);
    }
    let encoded: String = url::form_urlencoded::byte_serialize(trimmed.as_bytes()).collect();
    let url = format!(
        "https://rest.bandsintown.com/artists/{}/events?app_id={}",
        encoded, APP_ID
    );
    let client = match reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()
    {
        Ok(c) => c,
        Err(_) => return Ok(vec![]),
    };
    let resp = match client.get(&url).send().await {
        Ok(r) => r,
        Err(_) => return Ok(vec![]),
    };
    if !resp.status().is_success() {
        return Ok(vec![]);
    }
    let raw: serde_json::Value = match resp.json().await {
        Ok(v) => v,
        Err(_) => return Ok(vec![]),
    };
    let arr = match raw.as_array() {
        Some(a) => a,
        None => return Ok(vec![]),
    };
    let mut out: Vec<BandsintownEvent> = Vec::with_capacity(arr.len().min(20));
    for item in arr.iter().take(20) {
        let venue = item.get("venue").cloned().unwrap_or(serde_json::Value::Null);
        let lineup = item
            .get("lineup")
            .and_then(|v| v.as_array())
            .map(|a| a.iter().filter_map(|s| s.as_str().map(String::from)).collect())
            .unwrap_or_default();
        out.push(BandsintownEvent {
            datetime: item.get("datetime").and_then(|v| v.as_str()).unwrap_or("").to_string(),
            venue_name: venue.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string(),
            venue_city: venue.get("city").and_then(|v| v.as_str()).unwrap_or("").to_string(),
            venue_region: venue.get("region").and_then(|v| v.as_str()).unwrap_or("").to_string(),
            venue_country: venue.get("country").and_then(|v| v.as_str()).unwrap_or("").to_string(),
            url: item.get("url").and_then(|v| v.as_str()).unwrap_or("").to_string(),
            lineup,
        });
    }
    Ok(out)
}
